src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -888,7 +925,482 @@ - (void)test005_DoBDXTransferAllowUpdateRequest
+    sOTAProviderDelegate.queryImageHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+        MTROTASoftwareUpdateProviderClusterQueryImageParams * params, QueryImageCompletion completion) {
+        XCTAssertEqualObjects(nodeID, @(kDeviceId1));
+        XCTAssertEqual(controller, sController);
+
+        sOTAProviderDelegate.queryImageHandler = nil;
+        [sOTAProviderDelegate respondAvailableWithDelay:@(0)
+                                                    uri:imagePath1
+                                            updateToken:updateToken1
+                                        softwareVersion:kUpdatedSoftwareVersion_10
+                                  softwareVersionString:kUpdatedSoftwareVersionString_10
+                                             completion:completion];
+        [queryExpectation1 fulfill];
+    };
+
+    // Set up the BDX transfer begin, block query and transfer end handlers for device1 so it can go ahead with the BDX transfer
+    sOTAProviderDelegate.transferBeginHandler = ^(NSNumber * nodeID, MTRDeviceController * controller, NSString * fileDesignator,
+        NSNumber * offset, MTRStatusCompletion outerCompletion) {
+        XCTAssertEqualObjects(nodeID, @(kDeviceId1));
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(fileDesignator, imagePath1);
+        XCTAssertEqualObjects(offset, @(0));
+
+        readHandle = [NSFileHandle fileHandleForReadingAtPath:fileDesignator];
+        XCTAssertNotNil(readHandle);
+
+        NSError * endSeekError;
+        XCTAssertTrue([readHandle seekToEndReturningOffset:&imageSize error:&endSeekError]);
+        XCTAssertNil(endSeekError);
+
+        // Don't actually respond until the second requestor has queried us for
+        // an image to ensure BDX for device1 starts only after device2 has queried us once.
+
+        // Set up the query handler for device2 to return image available at imagePath2 but a busy delay of 30 secs
+        sOTAProviderDelegate.queryImageHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+            MTROTASoftwareUpdateProviderClusterQueryImageParams * params, QueryImageCompletion innerCompletion) {
+            XCTAssertEqualObjects(nodeID, @(kDeviceId2));
+            XCTAssertEqual(controller, sController);
+
+            [sOTAProviderDelegate respondAvailableWithDelay:@(busyDelay)
+                                                        uri:imagePath2
+                                                updateToken:updateToken2
+                                            softwareVersion:kUpdatedSoftwareVersion_5
+                                      softwareVersionString:kUpdatedSoftwareVersionString_5
+                                                 completion:innerCompletion];
+            if (!isQueryImageHandlerForDevice2Called) {
+                [queryExpectation2 fulfill];
+                sOTAProviderDelegate.transferBeginHandler = nil;
```

Comment:
The `sOTAProviderDelegate.transferBeginHandler = nil;` should stay in the outer bit, no?  Once the handler has been called once, we don't expect it to be called again until we respond that an image is actually available.

Reply:
yup it should. done.